# **Laboratorio 2**
**Daniela Navas**

## **Task 1** - Preguntas Teóricas

**1. ¿Por qué el modelo de Naive Bayes se le considera “naive”?**<br>
Se considera *ingenuo* debido a la suposición de que las características o predictores en un conjunto de datos son independientes entre sí, lo que se conoce como independencia condicional de clase. Esta suposición simplifica el cálculo de las probabilidades, ya que solo se necesita calcular la probabilidad de cada característica de forma individual, sin considerar las interacciones entre ellas. Aunque esta suposición no siempre se cumple en la realidad, donde las características suelen estar interrelacionadas, el algoritmo sigue siendo efectivo y fácil de implementar, especialmente con grandes conjuntos de datos.


**2. Explique la formulación matemática que se busca optimizar en Support Vector Machine, además responda ¿cómo funciona el truco del Kernel para este modelo? (Lo que se espera de esta pregunta es que puedan explicar en sus propias palabras la fórmula a la que llegamos que debemos optimizar de SVM en clase)**<br>
**Support Vector Machine**<br>
En Support Vector Machine (SVM), el objetivo es encontrar el hiperplano que mejor separa las clases en el espacio de características.<br> 
Primero se determina un hiperplano que separe dos clases de datos en un espacio de características. El hiperplano se define por la ecuación:

$$\mathbf{w} \cdot \mathbf{x} + b = 0$$

donde:
- Vector de pesos: $ \mathbf{w} $
- Sesgo (bias): $ b $
- Vector de características: $ \mathbf{x} $


Para que los datos estén correctamente clasificados, se deben cumplir las siguientes condiciones:
- Para puntos de la clase $+1$: $ \mathbf{w} \cdot \mathbf{x}_i + b \geq 1 $
- Para puntos de la clase $-1$: $ \mathbf{w} \cdot \mathbf{x}_i + b \leq -1 $

Combinandolas en una sola:

$$y_i (\mathbf{w} \cdot \mathbf{x}_i + b) \geq 1 \quad \forall i$$

donde $y_i$ es la etiqueta de la clase (+1 o -1).

El margen es la distancia entre los puntos más cercanos de cada clase y el hiperplano. Queremos maximizar este margen. La distancia de un punto al hiperplano es:

$$\frac{1}{\|\mathbf{w}\|}$$

Para maximizar el margen, minimizamos $\|\mathbf{w}\|$. Sin embargo, para simplificar los cálculos, minimizamos $\frac{1}{2} \|\mathbf{w}\|^2$ en su lugar.

El problema de optimización se formula como:

$$\min_{\mathbf{w}, b} \frac{1}{2} \|\mathbf{w}\|^2$$

sujeto a las restricciones:

$$y_i (\mathbf{w} \cdot \mathbf{x}_i + b) \geq 1 \quad \forall i$$

Esto nos lleva a un problema de optimización cuadrática con restricciones lineales. Se puede resolver usando métodos como los multiplicadores de Lagrange y técnicas de programación cuadrática.

Si los datos no son linealmente separables en el espacio original, utilizamos el truco del Kernel para mapear los datos a un espacio de mayor dimensión donde sí sean separables. Esto se hace mediante una función de kernel $K(\mathbf{x}_i, \mathbf{x}_j)$ que calcula el producto interno en el espacio transformado.

**Truco del Kernel**<br>
El truco del Kernel permite a SVM manejar datos que no son linealmente separables en el espacio original de características. La idea es transformar los datos a un espacio de mayor dimensión donde sí sean linealmente separables. Esto se hace mediante una función de kernel $K(\mathbf{x}_i, \mathbf{x}_j)$ que calcula el producto interno en el espacio transformado sin necesidad de calcular explícitamente la transformación.

Si hay un conjunto de datos $\{\mathbf{x}_i, y_i\}$ donde $\mathbf{x}_i$ es el vector de características y $y_i$ es la etiqueta de clase. En el espacio original, los datos no son linealmente separables.

Queremos transformar los datos a un espacio de mayor dimensión utilizando una función de transformación $\phi(\mathbf{x})$. Sin embargo, calcular explícitamente $\phi(\mathbf{x})$ puede ser computacionalmente costoso.

En lugar de calcular $\phi(\mathbf{x})$ directamente, utilizamos una función de kernel $K(\mathbf{x}_i, \mathbf{x}_j)$ que calcula el producto interno en el espacio transformado:

$$K(\mathbf{x}_i, \mathbf{x}_j) = \phi(\mathbf{x}_i) \cdot \phi(\mathbf{x}_j)$$


**3. Investigue sobre Random Forest y responda**<br>  
**a. ¿Qué tipo de ensemble learning es este modelo?**<br> 
Random Forest es un ejemplo de bagging (bootstrap aggregating). En este enfoque, se generan múltiples modelos (en este caso, árboles de decisión) entrenados en diferentes subconjuntos de datos obtenidos mediante muestreo con reemplazo. Este método busca reducir la varianza y mejorar la estabilidad del modelo al promediar las predicciones de varios árboles.

**b. ¿Cuál es la idea general detrás de Random Forest?**<br> 
Crear un conjunto de árboles de decisión que trabajan juntos para mejorar la precisión de las predicciones. Cada árbol es entrenado con una muestra aleatoria de los datos y, durante el proceso de construcción, se selecciona un subconjunto aleatorio de características para cada división en el árbol. Esto permite que los árboles sean diversos y, al final, sus resultados se combinan mediante votación (en clasificación) o promediación (en regresión) para obtener una predicción final más robusta.

**c. ¿Por qué se busca baja correlación entre los árboles de Random Forest?**<br>
La baja correlación entre los árboles en un modelo de Random Forest es crucial porque permite que cada árbol capture diferentes patrones en los datos. Si los árboles están muy correlacionados, sus errores también lo estarán, lo que puede llevar a una reducción limitada en la varianza del modelo. Al promover la diversidad entre los árboles, se mejora la capacidad del modelo para generalizar a nuevos datos y se reduce el riesgo de sobreajuste. Este enfoque asegura que la combinación de las predicciones sea más efectiva y robusta.

## **Task 2** -  Naive Bayes: Clasificador de Mensajes Ham/Spam<br>
Deberá construir un programa que reciba como entrada un archivo llamado “entrenamiento.txt” que será su dataset para entrenar un modelo basado en Bayes/Laplace Smoothing para clasificar mensajes como ham o spam. De dicho modelo, deberá reportar alguna métrica de desempeño. Además, con dicho modelo deberá ser capaz de interpretar mensajes futuros como spam o ham. Asimismo, deberá considerar las siguientes restricciones: 
- Solamente se podrá entrenar un modelo por dataset de entrenamiento (No se pueden cargar más de un archivo para entrenar el modelo) 
- Deberá limpiar el dataset de caracteres especiales y de combinaciones de mayúsculas/minúsculas. 
- Cada línea representa una observación (mensaje con su respectiva categoría) 
-  Del dataset dado deberá dividirlo en training y test. 
    - Para este punto, podrá usar librerías externas como las dadas en sklearn.  
    - **NO** se aceptará el uso de librerías para la construcción del modelo principal

### **Task 2.1** - Lectura y limpieza del dataset<br> 
Reciba como entrada un archivo llamado “entrenamiento.txt” que tendrá una estructura como la que se muestra 
abajo. En dicho archivo, cada línea representa un mensaje, y la primera palabra de la línea indica si es un mensaje 
Ham o Spam (etiqueta/categoría del mensaje), luego encuentra una tabulación (\t) para separar la etiqueta del 
verdadero mensaje, finalmente está el mensaje que termina hasta que encuentre un cambio de línea (\n). En base a 
este archivo, usted deberá entrenar un modelo basado en Bayes con Laplace Smoothing para clasificar mensajes 
como spam o ham.<br>  
Además, deberá limpiar el archivo de texto, teniendo en cuenta que pueden haber caracteres especiales que 
solamente agreguen ruido al mensaje y otros que pueden llegar a ser de utilidad para la clasificación. De igual modo, 
se recomienda cambiar los mensajes a que todos sigan la misma nomenclatura, es decir, todo en mayúsculas o bien, 
todo en minúscula, o similar, de modo tal que pueda clasificar de mejor manera la participación de cada palabra para 
determinar la categoría del mensaje.<br>   
También, deberá separar el dataset con un 80% para training, 20% para testing. Si llegan a necesitar una parte para 
validation, pueden subdividir el 20% de testing en 10% para validation y 10% para testing. Recuerden que esta división 
siempre deberá ser realizada de forma aleatoria, para reproducir sus resultados usen una seed.<br>  


In [1]:
import pandas as pd
import numpy as np
import re
from sklearn.model_selection import train_test_split

# Abrir archivo 
with open("entrenamiento.txt", "r", encoding="utf-8") as f:
    lines = f.readlines()

# Convertir en un set donde se tenga la etiqueta y el mensaje
data = []
for line in lines:
    parts = line.strip().split("\t", 1)  # Separar solo en la primera tabulación
    if len(parts) == 2:
        label, message = parts
        data.append((label, message)) # Nombrar "label" si es Spam o Ham; y "message" al contenido en si del mensaje

# Convertir a DataFrame
df = pd.DataFrame(data, columns=["label", "message"])

# Limpieza de datos | Eliminar ruido
def clean_text(text):
    text = text.lower()                      # Convertir a minúsculas (Normalizar)
    text = re.sub(r'[^a-z0-9\s]', '', text)  # Eliminar caracteres especiales (Filtro ruido)
    return text

df["message"] = df["message"].apply(clean_text) # Despues de separar, aplicar limpieza a todos los datos

# División en conjuntos de entrenamiento (80%) y prueba (20%)
train_data, test_data = train_test_split(df, test_size=0.2, random_state=42)


### **Task 2.2** - Construcción del modelo<br> 
Una vez tenga una forma de leer y limpiar el dataset, deberá crear un modelo basado en Bayes con Laplace Smoothing para que clasifique los mensajes en spam o ham. Esto deberá basarse en la probabilidad de cada palabra en pertenecer a cada uno de los posibles grupos. Cuando tenga la probabilidad del mensaje que sea spam o ham, deberá clasificarlo con la probabilidad de la categoría que resulte mayor.<br>  
Es importante tomar en cuenta que el entrenamiento y construcción del modelo deberá hacerse usando solamente la parte de training del dataset.<br>  

In [21]:
from collections import defaultdict
from sklearn.metrics import accuracy_score, confusion_matrix

# Entenamiento basado en probabilidad de Bayes 
def entrenar_bayes(mensajes, etiquetas, alpha=1):
    vocabulario = set()                                                     # Almacenar palabras únicas
    conteo_palabras = {'ham': defaultdict(int), 'spam': defaultdict(int)}   # Contar cuantas veces aparece cada palabra en spam y ham
    total_palabras = {'ham': 0, 'spam': 0}                                  # Total de palabras en spam y ham
    total_mensajes = {'ham': 0, 'spam': 0}                                  # Total de mensajes spam y ham
    
    for mensaje, etiqueta in zip(mensajes, etiquetas): # Recorrer cada mensaje y: 
        total_mensajes[etiqueta] += 1                  # Sumar en ham o spam, según corresponde
        palabras = mensaje.split()                     # Separar palabras 
        for palabra in palabras:                       # Contar cuantas veces aparece cada palabra en cada una de sus categorías. 
            vocabulario.add(palabra)
            conteo_palabras[etiqueta][palabra] += 1
            total_palabras[etiqueta] += 1              # Sumar en la cuenta general de palabras
    
    return vocabulario, conteo_palabras, total_palabras, total_mensajes #Retorna valores necesarios para el calculo de la probabilidad usando Bayes 

# Cálculo de Probabilidades usando Naive Bayes con Laplace Smoothing
def calcular_probabilidades(mensaje, vocabulario, conteo_palabras, total_palabras, total_mensajes, alpha=1): # Recibe valores del conteo que se hizo previamente 
    categorias = ['ham', 'spam']                         # Se tiene spam o ham
    total_mensajes_total = sum(total_mensajes.values())  
    probabilidades = {}
    palabras = mensaje.split()
    
    for categoria in categorias: # Recorrer Spam y ham
        prob_categoria = np.log(total_mensajes[categoria] / total_mensajes_total) # Proabilidad de uno u otro | Uso de log ya que al multiplicar probabilidades se obtienen cantidades pequeñas por lo que se toma el logaritmo de cada una y se suman, lo que evita que los valores se reduzcan demasiado.
        prob_palabras = 0 # Iniciar en cero
        for palabra in palabras: # Recorrer todas las palabras 
            conteo = conteo_palabras[categoria][palabra] + alpha # Laplace Smoothing para evitar probabilidades de cero
            prob_palabras += np.log(conteo / (total_palabras[categoria] + alpha * len(vocabulario)))
        probabilidades[categoria] = prob_categoria + prob_palabras 
    
    return max(probabilidades, key=probabilidades.get) # Retornar la categoría con mayor probabilidad (Spam o Ham)

def predecir(mensajes, vocabulario, conteo_palabras, total_palabras, total_mensajes): # Maximixar el proceso definiendo función que aplica calcular_probabilidades a cada mensaje para predecir su categoría
    return [calcular_probabilidades(mensaje, vocabulario, conteo_palabras, total_palabras, total_mensajes) for mensaje in mensajes]

# Entrenamiento de modelo usando la data de entrenamineto (train_data)
vocabulario, conteo_palabras, total_palabras, total_mensajes = entrenar_bayes(train_data["message"], train_data["label"])

Cuando ya tengan entrenado su modelo, deberán probar el modelo usando la parte de testing del dataset. La métrica a utilizar deberá proponerla considerando la distribución de clases/categorías en el dataset. Recuerde dejar justificada su respuesta en los comentarios de su código.<br>  
Presente al final del entrenamiento, la métrica de desempeño sobre el subset de training y sobre el subset de testing.<br>

In [22]:
# Evaluar el rendimiento del modelo usando la data de prueba (test_data) y lo obtenido del modelo entrenado
predicciones = predecir(test_data["message"], vocabulario, conteo_palabras, total_palabras, total_mensajes)

train_predicciones = predecir(train_data["message"], vocabulario, conteo_palabras, total_palabras, total_mensajes) # Luego tambien en train (Se pide en el enunciado)

# Obtención de métricas de rendimiento
# Matriz de confusión
labels = ['ham', 'spam']
cm = confusion_matrix(test_data["label"], predicciones, labels=labels)
TP = cm[1, 1]  # spam correctamente clasificado
FP = cm[0, 1]  # ham clasificado como spam
FN = cm[1, 0]  # spam clasificado como ham
TN = cm[0, 0]  # ham correctamente clasificado

# Calcular métricas
precision_metric = TP / (TP + FP) if (TP + FP) > 0 else 0
sensibilidad = TP / (TP + FN) if (TP + FN) > 0 else 0
especificidad = TN / (TN + FP) if (TN + FP) > 0 else 0

# TEST DATA
# Mostrar la matriz de confusión
print("\n--- TEST DATA ---")
print("Matriz de confusión:")
print(f"TP: {TP}, FP: {FP}")
print(f"FN: {FN}, TN: {TN}")
print("------------------------------------------")

# Métrica de desempeño: Precisión, Sensibilidad y Especificidad
print("Precisión:", precision_metric*100, "%")
print("Sensibilidad:", sensibilidad*100, " %")
print("Especificidad:", especificidad*100, " %")


# TRAIN DATA
train_precision = accuracy_score(train_data["label"], train_predicciones)
train_cm = confusion_matrix(train_data["label"], train_predicciones, labels=['ham', 'spam'])
train_TP = train_cm[1, 1]
train_FP = train_cm[0, 1]
train_FN = train_cm[1, 0]
train_TN = train_cm[0, 0]
train_precision_metric = train_TP / (train_TP + train_FP) if (train_TP + train_FP) > 0 else 0
train_sensibilidad = train_TP / (train_TP + train_FN) if (train_TP + train_FN) > 0 else 0
train_especificidad = train_TN / (train_TN + train_FP) if (train_TN + train_FP) > 0 else 0

# Mostrar la matriz de confusión
print("\n--- TRAIN DATA ---")
print("Matriz de confusión:")
print(f"TP: {train_TP}, FP: {train_FP}")
print(f"FN: {train_FN}, TN: {train_TN}")
print("------------------------------------------")

# Métrica de desempeño: Precisión, Sensibilidad y Especificidad
print("Precisión:", train_precision_metric*100, "%")
print("Sensibilidad:", train_sensibilidad*100, " %")
print("Especificidad:", train_especificidad*100, " %")



--- TEST DATA ---
Matriz de confusión:
TP: 157, FP: 18
FN: 6, TN: 932
------------------------------------------
Precisión: 89.71428571428571 %
Sensibilidad: 96.31901840490798  %
Especificidad: 98.10526315789474  %

--- TRAIN DATA ---
Matriz de confusión:
TP: 559, FP: 14
FN: 25, TN: 3854
------------------------------------------
Precisión: 97.55671902268762 %
Sensibilidad: 95.71917808219177  %
Especificidad: 99.63805584281282  %


Por medio de ello, se puede definir que el modelo sobre el subset de training tiene:<br>
- **Precisión:** 97.57% <br>
- **Sensibilidad:** 95.72% <br>
- **Especificidad:** 99.64% <br>

Y el modelo sobre el subset de testing tiene:<br>
- **Precisión:** 89.72% <br>
- **Sensibilidad:** 96.31% <br>
- **Especificidad:** 98.11% <br>

El modelo muestra un excelente desempeño en el conjunto de entrenamiento, con una precisión del 97.57%, una alta sensibilidad del 95.72% y una especificidad del 99.64%. Esto indica que el modelo clasifica correctamente la mayoría de los correos, tanto spam como ham, con un bajo índice de falsos positivos y falsos negativos. Sin embargo, en el conjunto de prueba, la precisión disminuye a 89.72%, lo que sugiere una menor capacidad para clasificar correctamente los correos en datos no vistos. A pesar de ello, la sensibilidad sigue siendo alta (96.31%), lo que indica que el modelo sigue detectando correctamente la mayoría de los correos spam, aunque la ligera disminución en la especificidad (98.11%) implica un pequeño aumento en los falsos positivos. Esta diferencia de rendimiento entre los conjuntos de entrenamiento y prueba sugiere que el modelo puede estar ligeramente sobreajustado, es decir, que se ha ajustado demasiado a los datos de entrenamiento y ha perdido algo de generalización. Aún así, su rendimiento general sigue siendo sólido, especialmente en la detección de correos spam.

---

### **Task 2.3** - Clasificación de mensajes futuros<br> 
Al finalizar el entrenamiento de su modelo, permita que se ingresen nuevos mensajes de forma individual desde su interfaz (ya sea en consola o en una interfaz gráfica). Y que estos se clasifiquen usando su modelo. Su programa deberá poder recibir cualquier tipo de mensaje y devolver la probabilidad de que sea spam y ham, así mismo como cuál es la decisión de clasificación de su modelo.  

In [28]:
# Función para clasificar un mensaje ingresado por el usuario
def clasificar_mensaje():
    while True: # Usar un ciclo While
        mensaje = input("Ingrese un mensaje para clasificar (o 'salir' para terminar): ") # Leer entrada
        if mensaje.lower() == 'salir': # Si es salir
            print("SALIENDO... \n")
            break                      # Terminar ejecución
        print("El mensaje: " , mensaje)
        mensaje = clean_text(mensaje) 
        clasificacion = calcular_probabilidades(mensaje, vocabulario, conteo_palabras, total_palabras, total_mensajes)
        print(f"Ha sido clasificado como: {clasificacion}\n")
    
clasificar_mensaje()


El mensaje:  hello, how are you?
Ha sido clasificado como: ham

El mensaje:  Yup... Ok i go home look at the timings then i msg ü again... Xuhui going to learn on 2nd may too but her lesson is at 8am
Ha sido clasificado como: ham

El mensaje:  XXXMobileMovieClub: To use your credit, click the WAP link in the next txt message or click here>> http://wap. xxxmobilemovieclub.com?n=QJKGIGHJJGCBL
Ha sido clasificado como: spam

El mensaje:  Yup... Ok i go home look at the timings then i msg ü again... Xuhui going to learn on 2nd may too but her lesson is at 8am
Ha sido clasificado como: ham

El mensaje:  Yup... Ok i go home look at the timings then i msg ü again... Xuhui going to learn on 2nd may too but her lesson is at 8am
Ha sido clasificado como: ham

SALIENDO... 



### **Task 2.4** - Comparación con Librerías<br> 
Entrene un modelo de Bayes usando alguna librería ya existente como sklearn.naive_bayes.MultinomialNB en Python. Este deberá usar el mismo dataset dado, y deberá además dividirlo en training y testing como en el Task 2.1. Compare la métrica de desempeño tanto en training como en testing con la misma implementación que usted realizó.

In [30]:
import pandas as pd
import numpy as np
import re
from sklearn.model_selection import train_test_split
from sklearn.feature_extraction.text import CountVectorizer
from sklearn.naive_bayes import MultinomialNB
from sklearn.metrics import confusion_matrix, precision_score, recall_score, accuracy_score

# Abrir archivo 
with open("entrenamiento.txt", "r", encoding="utf-8") as f:
    lines = f.readlines()

# Convertir en un set donde se tenga la etiqueta y el mensaje
data = []
for line in lines:
    parts = line.strip().split("\t", 1)  # Separar solo en la primera tabulación
    if len(parts) == 2:
        label, message = parts
        data.append((label, message)) # Nombrar "label" si es Spam o Ham; y "message" al contenido en si del mensaje

# Convertir a DataFrame
df = pd.DataFrame(data, columns=["label", "message"])

# Limpieza de datos | Eliminar ruido
def clean_text(text):
    text = text.lower()                      # Convertir a minúsculas (Normalizar)
    text = re.sub(r'[^a-z0-9\s]', '', text)  # Eliminar caracteres especiales (Filtro ruido)
    return text

df["message"] = df["message"].apply(clean_text) # Despues de separar, aplicar limpieza a todos los datos

# División en conjuntos de entrenamiento (80%) y prueba (20%)
train_data, test_data = train_test_split(df, test_size=0.2, random_state=42)

# Vectorización de los mensajes (transformar texto en vectores numéricos)
vectorizer = CountVectorizer()
X_train = vectorizer.fit_transform(train_data['message'])
X_test = vectorizer.transform(test_data['message'])

# Etiquetas (Spam o Ham)
y_train = train_data['label']
y_test = test_data['label']

# Modelo de Naive Bayes
model = MultinomialNB()
model.fit(X_train, y_train)

# Predicciones en el conjunto de prueba
y_pred_train = model.predict(X_train)
y_pred_test = model.predict(X_test)

# Métricas de desempeño
# TRAIN DATA
conf_matrix_train = confusion_matrix(y_train, y_pred_train)

TP_train = conf_matrix_train[1, 1]  # spam correctamente clasificado
FP_train = conf_matrix_train[0, 1]  # ham clasificado como spam
FN_train = conf_matrix_train[1, 0]  # spam clasificado como ham
TN_train = conf_matrix_train[0, 0]  # ham correctamente clasificado

precision_train = precision_score(y_train, y_pred_train, pos_label='spam')
recall_train = recall_score(y_train, y_pred_train, pos_label='spam')
specificity_train = conf_matrix_train[1,1] / (conf_matrix_train[1,1] + conf_matrix_train[0,1])

# Mostrar la matriz de confusión
print("\n--- TRAIN DATA ---")
print("Matriz de confusión:")
print(f"TP: {TP_train}, FP: {FP_train}")
print(f"FN: {FN_train}, TN: {TN_train}")
print("------------------------------------------")

# Métrica de desempeño: Precisión, Sensibilidad y Especificidad
print("Precisión:", precision_train*100, "%")
print("Sensibilidad:", recall_train*100, " %")
print("Especificidad:", specificity_train*100, " %")

# ---------------------------------------------------------------------------------------------------------------------------
# TEST DATA
conf_matrix_test = confusion_matrix(y_test, y_pred_test)

TP_test = conf_matrix_test[1, 1]  # spam correctamente clasificado
FP_test = conf_matrix_test[0, 1]  # ham clasificado como spam
FN_test = conf_matrix_test[1, 0]  # spam clasificado como ham
TN_test = conf_matrix_test[0, 0]  # ham correctamente clasificado

precision_test = precision_score(y_test, y_pred_test, pos_label='spam')
recall_test = recall_score(y_test, y_pred_test, pos_label='spam')
specificity_test = conf_matrix_test[1,1] / (conf_matrix_test[1,1] + conf_matrix_test[0,1])

# Mostrar la matriz de confusión
print("\n--- TEST DATA ---")
print("Matriz de confusión:")
print(f"TP: {TP_test}, FP: {FP_test}")
print(f"FN: {FN_test}, TN: {TN_test}")
print("------------------------------------------")

# Métrica de desempeño: Precisión, Sensibilidad y Especificidad
print("Precisión:", precision_test*100, "%")
print("Sensibilidad:", recall_test*100, " %")
print("Especificidad:", specificity_test*100, " %")


--- TRAIN DATA ---
Matriz de confusión:
TP: 560, FP: 14
FN: 24, TN: 3854
------------------------------------------
Precisión: 97.5609756097561 %
Sensibilidad: 95.8904109589041  %
Especificidad: 97.5609756097561  %

--- TEST DATA ---
Matriz de confusión:
TP: 151, FP: 4
FN: 12, TN: 946
------------------------------------------
Precisión: 97.41935483870968 %
Sensibilidad: 92.63803680981594  %
Especificidad: 97.41935483870968  %


Por medio de ello, se puede definir que el modelo sobre el subset de training tiene:<br>
- **Precisión:** 97.57% <br>
- **Sensibilidad:** 95.89% <br>
- **Especificidad:** 97.56% <br>

Y el modelo sobre el subset de testing tiene:<br>
- **Precisión:** 87.42% <br>
- **Sensibilidad:** 92.63% <br>
- **Especificidad:** 97.42% <br>

**¿Cuál implementación lo hizo mejor? ¿Su implementación o la de la librería?**
Comparando ambos modelos, podemos observar que ambos tienen un desempeño similar en el conjunto de entrenamiento, con una precisión idéntica del 97.57%. Sin embargo, al analizar el conjunto de prueba, el modelo hecho a mano tiene una mayor precisión (89.72%) en comparación con el modelo implementado con librerías (87.42%). El modelo hecho a mano muestra un mejor rendimiento en el conjunto de prueba, lo que podría indicar que está mejor ajustado para generalizar a datos no vistos. Sin embargo, esto no siempre es ideal. Un modelo que se ajusta demasiado a los datos de entrenamiento puede perder capacidad de generalización, y aunque la precisión sea mayor, es posible que no se mantenga ese buen rendimiento en situaciones más complejas o con datos adicionales.


**¿Por qué cree que se debe esta diferencia?**
Las librerías de machine learning, como las que se usan en bibliotecas como `scikit-learn`, suelen aplicar optimizaciones y técnicas de regularización que ayudan a evitar el sobreajuste, pero a costa de una ligera reducción en la precisión, especialmente en el conjunto de prueba. Aunque los resultados del modelo con librerías muestran una disminución en la precisión, su capacidad para evitar sobreajuste podría ser más efectiva a largo plazo, garantizando que el modelo no aprenda patrones espurios específicos de los datos de entrenamiento. Las librerías suelen incluir técnicas de preprocesamiento más robustas, manejo de datos faltantes, normalización, y validación cruzada que pueden ayudar a mejorar el rendimiento general. El modelo hecho a mano podría no haber utilizado todas estas técnicas o podría haber tenido limitaciones en su implementación.